src/rust/engine/fs/src/snapshot.rs

```diff
@@ -496,6 +498,36 @@ impl Snapshot {
       })
       .to_boxed()
   }
+
+  pub fn from_files_content(
+    store: Store,
```

Comment:
Solely curiousity - why not use an address `&Store`? Also I see in the final line we move the `store` variable into `Snapshot::from_path_stats`, which means that that function becomes the new owner, right?

Reply:
Yep! `Store` is a tiny type that cheaply implements `Clone` by wrapping something that is threadsafe and reference counted.

We do a lot of this currently due to Future not making borrowing very ergonomic. See http://aturon.github.io/2018/04/24/async-borrowing/

Reply:
> Store is a tiny type that cheaply implements Clone

Ah cool, just like why you don't use references / borrowing for `i32`. Thanks!

Reply:
> > Store is a tiny type that cheaply implements Clone
> 
> Ah cool, just like why you don't use references / borrowing for `i32`. Thanks!

There's a little more to this, actually - ideally we _would_ use references and borrowing here, because cloning this isn't nearly as cheap as copying an `i32`.

One of the functions we're going to end up passing this `Store` to is going to end up moving it onto another thread (to use it from a Future). If we borrowed a `Store` here, we'd need to clone the `Store` at the point that we move it onto that thread (because you can't borrow across that boundary).

As a rule, if you're going to borrow a value, and then unconditionally clone it, it's better to just take an owned value in the first place, because that's being up-front about the cost of the function. That said, when async/await happens, we can probably get rid of a bunch of those clones, because we should be able to borrow across futures boundaries.